frontend/src/components/dashboard/hooks/useGrouping.ts

```diff
@@ -1,195 +1,95 @@
-  category: Category;
-  solutions: AnalyzedSolution[];
-};
-
-export type Group = {
-  key: string;
-  label: string;
-
-  minX: number;
-  maxX: number;
-
-  minY: number;
-  maxY: number;
-};
-
-const charCodeOfCapitalA = 65;
-const numberOfCharacters = 26;
-
-const getGroupName = (idx: number): string => {
-  // converts the number to base 26 with 0 = A, 1 = B, ..., 25 = Z, 26 = BA,...
-  const name = [];
-
-  do {
-    name.push(
-      String.fromCharCode(charCodeOfCapitalA + (idx % numberOfCharacters)),
-    );
-
-    idx = Math.floor(idx / numberOfCharacters);
-  } while (idx > 0);
-
-  return name.reverse().join("");
-};
-
-const isWithinGroup = (
-  group: Omit<Group, "label">,
-  x: number,
-  y: number,
-): boolean =>
-  group.minX <= x && x < group.maxX && group.minY <= y && y < group.maxY;
 
 export const useGrouping = (
+  isAutomaticGrouping: boolean,
+  numberOfGroups: number,
   solutions: CurrentAnalysis[] | undefined,
   filters: FilterCriterion[],
   splits: ChartSplit[],
   xAxis: AxesCriterionType,
   yAxis: AxesCriterionType,
 ): {
-  categoriesWithGroups: CategoryWithGroups[];
+  categorizedDataPoints: CategorizedDataPoints[];
+  groupAssignments: SolutionGroupAssignment[];
   groups: Group[];
-} =>
-  useMemo(() => {
-    if (!solutions) {
-      return {
-        categoriesWithGroups: [],
-        groups: [],
-      };
-    }
-
-    const groups: (Omit<Group, "label"> & { label?: string })[] = [];
-
-    const verticalSplits = splits
-      .filter((s) => s.type === SplitType.vertical)
-      .sort((a, b) => a.x - b.x);
-    const horizontalSplits = splits
-      .filter((s) => s.type === SplitType.horizontal)
-      .sort((a, b) => a.y - b.y);
-
-    let lastX = Number.NEGATIVE_INFINITY;
-
-    for (const vSplit of [
-      ...verticalSplits,
-      // add an artificial split to include the groups after the last vertical split
-      { type: SplitType.vertical, x: Number.POSITIVE_INFINITY },
-    ]) {
-      let lastY = Number.NEGATIVE_INFINITY;
-      for (const hSplit of [
-        ...horizontalSplits,
-        // add an artificial split to include the groups after the last horizontal split
-        {
-          type: SplitType.horizontal,
-          y: Number.POSITIVE_INFINITY,
+  manualGroups: ManualGroup[];
+} => {
+  const [dataPoints, setDataPoints] = useState<CategorizedDataPoints[]>([]);
+  const [groupAssignment, setGroupAssignment] = useState<
+    SolutionGroupAssignment[]
+  >([]);
+  const [groups, setGroups] = useState<Group[]>([]);
+  const [manualGroups, setManualGroups] = useState<ManualGroup[]>([]);
+
+  const manual = useManualGrouping(
+    isAutomaticGrouping,
+    solutions,
+    filters,
+    splits,
+    xAxis,
+    yAxis,
+  );
+
+  const computeAutomaticGrouping = useAutomaticGrouping(
+    isAutomaticGrouping,
+    numberOfGroups,
+    solutions,
+    filters,
+    xAxis,
+    yAxis,
+  );
+
+  useEffect(() => {
+    let isCancelled = false;
+    if (isAutomaticGrouping) {
```

Comment:
given this clause is much longer, consider starting with manual for readability, i.e.
```
if(!isAutomaticGrouping) {
      setDataPoints(manual.dataPoints);
      setGroupAssignment(manual.groupAssignment);
      setGroups(manual.groups);
      setManualGroups(manual.groups);
} else {
      // the much longer thing
}
```